appsmithorg/appsmith
# task: 12949 Trim unnecessary re-renders in baseWidget layers

## Description

Target:
1. Remove unnecessary re-renders in onion layers of the base widget.
2. Remove unnecessary renders in sibling widgets upon selection of a widget.

Changes:
1. Create selectors for checking if a widget is selected or focused.
2. Memoize WidgetName component.

Fixes #12949

## Type of change


## How Has This Been Tested?

1. Clone the A force app.
2. Click on an issue. Wait for the description to load. Click on the title input.
3. Turn on the React Profiler.
4.  Click on Title input and type a character.
5. Turn off the React profiler
6. Observe that there are no re-renders in the onion layers of other components.

Repeat steps 2-6 on the aforce app on release to observe the difference in the number of re-renders.

Repeat steps 2-6 on the release aforce and the cloned app on this branch. Profile using the performance tab. Observe the difference in the scripting times in the two scenarios.


## Checklist:











## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>


    // Code coverage diff between base branch:release and head branch: task/12949 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 56.62 **(0.01)** | 38.54 **(0)** | 35.89 **(0)** | 56.86 **(0.01)**
 :red_circle: | app/client/src/components/editorComponents/ResizableComponent.tsx | 63.1 **(-1.27)** | 21.88 **(2.21)** | 44.44 **(-13.89)** | 63.1 **(-1.27)**
 :red_circle: | app/client/src/components/editorComponents/SnipeableComponent.tsx | 77.78 **(0)** | 27.27 **(0)** | 20 **(-13.33)** | 77.78 **(0)**
 :red_circle: | app/client/src/components/editorComponents/WidgetNameComponent/index.tsx | 96.83 **(-0.14)** | 82.69 **(-0.95)** | 100 **(0)** | 96.72 **(-0.16)**
 :red_circle: | app/client/src/selectors/propertyPaneSelectors.tsx | 84.72 **(-1.39)** | 64.81 **(0)** | 83.33 **(-8.34)** | 87.1 **(-1.61)**
 :green_circle: | app/client/src/selectors/ui.tsx | 80.95 **(3.17)** | 100 **(0)** | 42.86 **(9.53)** | 71.43 **(4.76)**
 :green_circle: | app/client/src/selectors/widgetSelectors.ts | 73.21 **(10.71)** | 33.33 **(-2.38)** | 82.35 **(25.21)** | 71.43 **(9.53)**
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.94 **(0.23)** | 41.67 **(0.84)** | 36.21 **(0)** | 56.99 **(0.25)**
 :green_circle: | app/client/src/utils/hooks/useClickToSelectWidget.tsx | 64.81 **(1.04)** | 64 **(15.11)** | 40 **(-10)** | 62.75 **(0.63)**
 :green_circle: | app/client/src/widgets/BaseWidget.tsx | 80.3 **(0.46)** | 48.94 **(0)** | 77.14 **(0.67)** | 79.53 **(0.5)**</details>
